src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,141 @@
+            add(value);
+            return;
+        }
+        isIndexValid(index);
+        Node<T> node = getNode(index);
+        Node<T> newNode = new Node<>(node.prev, value, node);
+        if (node.prev == null) {
+            head = newNode;
+        } else {
+            node.prev.next = newNode;
+        }
+        node.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexValid(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isIndexValid(index);
+        Node<T> currentNode = getNode(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = getNode(index);
+        unlink(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> node = head; node != null; node = node.next) {
+            if (node.value == object || node.value != null && node.value.equals(object)) {
+                unlink(node);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void isIndexValid(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Can't get value by index: " + index);
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        isIndexValid(index);
+        Node<T> currentNode = head;
+        if (index < (size / 2)) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private void unlink(Node<T> removenode) {
+        if (removenode.prev == null) {
+            head = removenode.next;
+        } else {
+            removenode.prev.next = removenode.next;
+        }
+        if (removenode.next == null) {
+            tail = removenode.prev;
+        } else {
+            removenode.next.prev = removenode.prev;
+        }
+        size--;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(T value) {
+            this.value = value;
+        }
+
+        public Node(Node<T> prev, T value, Node<T> next) {
```

Comment:
```suggestion
        Node(T value) {
            this.value = value;
        }

        Node(Node<T> prev, T value, Node<T> next) {
```
we won't be able to create a node outside of this class anyway, so better be upfront about it and avoid possible confusion